driver/main.c

```diff
@@ -296,6 +302,189 @@ static void compat_unregister_trace(void *func, const char *probename, struct tr
+			continue;
+		}
+
+		switch(idx)
+		{
+		case SYS_ENTER:
+			if(new_val)
+			{
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 20)
+				ret = compat_register_trace(syscall_enter_probe, tp_names[idx], tp_sys_enter);
+#else
+				ret = register_trace_syscall_enter(syscall_enter_probe);
+#endif
+			}
+			else
+			{
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 20)
+				compat_unregister_trace(syscall_enter_probe, tp_names[idx], tp_sys_enter);
+#else
+				unregister_trace_syscall_enter(syscall_enter_probe);
+#endif
+			}
+			break;
+		case SYS_EXIT:
+			if(new_val)
+			{
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 20)
+				ret = compat_register_trace(syscall_exit_probe, tp_names[idx], tp_sys_exit);
+#else
+				ret = register_trace_syscall_exit(syscall_exit_probe);
+#endif
+			}
+			else
+			{
+#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 20)
+				compat_unregister_trace(syscall_exit_probe, tp_names[idx], tp_sys_exit);
+#else
+				unregister_trace_syscall_exit(syscall_exit_probe);
+#endif
+			}
+			break;
+		case SCHED_PROC_EXIT:
+			ret = compat_set_tracepoint(syscall_procexit_probe, tp_names[idx], tp_sched_process_exit, new_val);
+			break;
+#ifdef CAPTURE_CONTEXT_SWITCHES
+		case SCHED_SWITCH:
+			ret = compat_set_tracepoint(sched_switch_probe, tp_names[idx], tp_sched_switch, new_val);
+			break;
+#endif
+#ifdef CAPTURE_PAGE_FAULTS
+		case PAGE_FAULT_USER:
+			if (!g_fault_tracepoint_disabled)
+			{
+				ret = compat_set_tracepoint(page_fault_user_probe, tp_names[idx], tp_page_fault_user, new_val);
+			}
+			break;
+		case PAGE_FAULT_KERN:
+			if (!g_fault_tracepoint_disabled)
+			{
+				ret = compat_set_tracepoint(page_fault_kern_probe, tp_names[idx], tp_page_fault_kernel, new_val);
+			}
+			break;
+#endif
+#ifdef CAPTURE_SIGNAL_DELIVERIES
+		case SIGNAL_DELIVER:
+			ret = compat_set_tracepoint(signal_deliver_probe, tp_names[idx], tp_signal_deliver, new_val);
+			break;
+#endif
+#ifdef CAPTURE_SCHED_PROC_FORK
+		case SCHED_PROC_FORK:
+			ret = compat_set_tracepoint(sched_proc_fork_probe, tp_names[idx], tp_sched_proc_fork, new_val);
+			break;
+#endif
+#ifdef CAPTURE_SCHED_PROC_EXEC
+		case SCHED_PROC_EXEC:
+			ret = compat_set_tracepoint(sched_proc_exec_probe, tp_names[idx], tp_sched_proc_exec, new_val);
+			break;
+#endif
+		default:
+			// unmanaged idx
+			break;
+		}
+
+		if (new_val)
+		{
+			if (ret == 0)
+			{
+				g_tracepoints_attached |= 1 << idx;
+			}
+			else
+			{
+				pr_err("can't attach the %s tracepoint\n", tp_names[idx]);
+			}
+		}
+		else
+		{
+			if (ret == 0)
+			{
+				g_tracepoints_attached &= ~(1 << idx);
+			}
+			else
+			{
+				pr_err("can't detach the %s tracepoint\n", tp_names[idx]);
+			}
+		}
+	}
+
+	if (ret != 0)
+	{
+		// Error: reset first idx-1 bits to their old value.
+		// This means that we are requesting to reset first
+		// idx-1 tracepoints, that are the succedeed ones before the error.
+		force_tp_set(consumer, stored_tp_set, idx - 1);
+	}
+
+	if (g_tracepoints_attached == 0)
```

Comment:
This was done when we detached every tracepoint after last consumer was gone.
We now do it as soon as we haven't got any attached tracepoint.